EmptyAiaResponseIsIgnored should use incomplete chain.

vcsjones opened the pull request:
This changes the test to build a partial chain using two intermediates.
The leaf's issuing intermediate is added to the ExtraStore and the
second intermediate is expected to be fetched by AIA (and fail with
an empty response).

This is to help coerce chain builders treat it as a partial chain,
and can actually make some progress, as opposed to the leaf being
treated as an invalid root.

Fixes #48223

---------

Easier to review with "Hide whitespace changes".

vcsjones committed:
EmptyAiaResponseIsIgnored should use incomplete chain.

This changes the test to build a partial chain using two intermediates.
The leaf's issuing intermediate is added to the ExtraStore and the
second intermediate is expected to be fetched by AIA (and fail with
an empty response).

This is to help coerce chain builders treat it as a partial chain,
and can actually make some progress, as opposed to the leaf being
treated as an invalid root.

bartonjs commented on src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/AiaTests.cs:
```diff
@@ -16,21 +16,26 @@ public static void EmptyAiaResponseIsIgnored()
             using (root)
-            using (intermediate)
+            using (intermediate1)
+            using (intermediate2)
```
Why not use the variable declaration using for these?

vcsjones replied:
Uhhhhhhhh, because I... didn't. Will fix when I see what's up with Windows.

(I was sure I could just get this right on the first try from my Linux box..)

vcsjones committed:
Fix certificate cache resuse.

Since the certificate names were the same, different tests were
colliding with caches since they had the same certificate name.

vcsjones commented:
Okay, I think I figured it out. Windows was getting confused by certs with colliding names being in some kind of cache from different tests. The test always passed in isolation but was unhappy when run as part of the suite.

bartonjs commented:
> The test always passed in isolation but was unhappy when run as part of the suite.

Presumably that means that this test generated a CA with too much "sameness" to one that got fetched via AIA.  Do we need the longer chain, or just to reduce our sameness?

It looks like the call to BuildPrivatePki isn't specifying a value for test name (and it's not painted as `[CallerMemberName]`), so it's ending up with a subject name collision with other tests that similarly don't specify the name.  Probably one AIA test is caching while another is depending on no cache.  I think the fix is as simple as providing a value for testName.

The easiest way is to call DynamicRevocationTests.BuildPrivatePki instead of CertificateAuthority.BuildPrivatePki.  (The former calls the latter, but the former has `[CallerMemberName]` and the latter does not.)

vcsjones commented:
@bartonjs

> It looks like the call to BuildPrivatePki isn't specifying a value for test name

Well, that's what the second commit adds.

bartonjs commented:
Fair, fair.  I guess my question is "so do we need the dynamically longer chain logic, then?"  I guess it doesn't hurt.

vcsjones commented:
> "so do we need the dynamically longer chain logic, then?"

Ah. I see. I thought about it and decided at the very least it make what was going on much more obvious to me, and, might as well keep the code that lets you make N number of intermediates.
